src/System.Security.Cryptography.Algorithms/tests/AesCcmTests.cs

```diff
@@ -0,0 +1,602 @@
+        }
+
+        [ConditionalTheory(nameof(SupportsCcm))]
+        [MemberData(nameof(GetNistCcmTestCases))]
+        public static void AesCcmNistTests(AEADTest testCase)
+        {
+            using (var aesCcm = new AesCcm(testCase.Key))
+            {
+                byte[] ciphertext = new byte[testCase.Plaintext.Length];
+                byte[] tag = new byte[testCase.Tag.Length];
+                aesCcm.Encrypt(testCase.Nonce, testCase.Plaintext, ciphertext, tag, testCase.AssociatedData);
+                Assert.Equal(testCase.Ciphertext, ciphertext);
+                Assert.Equal(testCase.Tag, tag);
+
+                byte[] plaintext = new byte[testCase.Plaintext.Length];
+                aesCcm.Decrypt(testCase.Nonce, ciphertext, tag, plaintext, testCase.AssociatedData);
+                Assert.Equal(testCase.Plaintext, plaintext);
+            }
+        }
+
+        [ConditionalTheory(nameof(SupportsCcm))]
+        [MemberData(nameof(GetNistCcmTestCases))]
+        public static void AesCcmNistTestsTamperTag(AEADTest testCase)
+        {
+            using (var aesCcm = new AesCcm(testCase.Key))
+            {
+                byte[] ciphertext = new byte[testCase.Plaintext.Length];
+                byte[] tag = new byte[testCase.Tag.Length];
+                aesCcm.Encrypt(testCase.Nonce, testCase.Plaintext, ciphertext, tag, testCase.AssociatedData);
+                Assert.Equal(testCase.Ciphertext, ciphertext);
+                Assert.Equal(testCase.Tag, tag);
+
+                tag[0] ^= 1;
+
+                byte[] plaintext = new byte[testCase.Plaintext.Length];
+                Assert.Throws<CryptographicException>(
+                    () => aesCcm.Decrypt(testCase.Nonce, ciphertext, tag, plaintext, testCase.AssociatedData));
+            }
+        }
+
+        private static IEnumerable<object[]> GetValidNonceSizes()
+        {
+            return GetValidSizes(AesCcm.NonceByteSizes);
+        }
+
+        private static IEnumerable<object[]> GetInvalidNonceSizes()
+        {
+            return GetInvalidSizes(AesCcm.NonceByteSizes);
+        }
+
+        private static IEnumerable<object[]> GetValidTagSizes()
+        {
+            return GetValidSizes(AesCcm.TagByteSizes);
+        }
+
+        private static IEnumerable<object[]> GetInvalidTagSizes()
+        {
+            return GetInvalidSizes(AesCcm.TagByteSizes);
+        }
+
+        public static IEnumerable<object[]> GetNistCcmTestCases()
+        {
+            foreach (AEADTest test in s_nistCcmTestVectorsSelectedCases)
+            {
+                yield return new object[] { test };
+            }
+        }
+
+        // https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/mac/ccmtestvectors.zip
+        private const string NistCcmTestVectors = "NIST CCM Test Vectors";
+
+        private static readonly AEADTest[] s_nistCcmTestVectorsSelectedCases = new AEADTest[]
+        {
+            new AEADTest
```

Comment:
It took me a minute to understand why this case was interesting (aside from being the first one).

Maybe it would be useful to explicitly show the `Plaintext = null,` `Ciphertext = null`, and `AssociatedData = null,` values.  Otherwise you're looking for absent data.

Reply:
gaps filled